man/zabbix_agent2.man
.\" Start URL macro. Copied from an-ext.tmac for portability
.de UR
.  ds m1 \\$1\"
.  nh
.  if \\n(mH \{\
.    \" Start diversion in a new environment.
.    do ev URL-div
.    do di URL-div
.  \}
..
.\" End URL macro. Copied from an-ext.tmac for portability
.de UE
.  ie \\n(mH \{\
.    br
.    di
.    ev
.
.    \" Has there been one or more input lines for the link text?
.    ie \\n(dn \{\
.      do HTML-NS "<a href=""\\*(m1"">"
.      \" Yes, strip off final newline of diversion and emit it.
.      do chop URL-div
.      do URL-div
\c
.      do HTML-NS </a>
.    \}
.    el \
.      do HTML-NS "<a href=""\\*(m1"">\\*(m1</a>"
\&\\$*\"
.  \}
.  el \
\\*(la\\*(m1\\*(ra\\$*\"
.
.  hy \\n(HY
..
.TH ZABBIX_AGENT2 8 "2019\-01\-29" Zabbix
.SH NAME
zabbix_agent2 \- Zabbix agent 2
.SH SYNOPSIS
.B zabbix_agent2
.RB [ \-c
.IR config\-file ]
.br
.B zabbix_agent2
.RB [ \-c
.IR config\-file ]
.B \-p
.br
.B zabbix_agent2
.RB [ \-c
.IR config\-file ]
.B \-t
.I item\-key
.br
.B zabbix_agent2
.RB [ \-c
.IR config\-file ]
.B \-R
.I runtime\-option
.br
.B zabbix_agent2 \-h
.br
.B zabbix_agent2 \-V
.SH DESCRIPTION
.B zabbix_agent2
is an application for monitoring parameters of various services. 
.SH OPTIONS
.IP "\fB\-c\fR, \fB\-\-config\fR \fIconfig\-file\fR"
Use the alternate \fIconfig\-file\fR instead of the default one.
.IP "\fB\-R\fR, \fB\-\-runtime\-control\fR \fIruntime\-option\fR"
Perform administrative functions according to \fIruntime\-option\fR.
.SS
.RS 4
Runtime control options
.RS 4
.TP 4
\fBloglevel increase\f
Increase log level
.RE
.RS 4
.TP 4
\fBloglevel decrease\f
Decrease log level
.RE
.RS 4
.TP 4
\fBhelp\f
List available runtime control options
.RE
.RS 4
.TP 4
\fBmetrics\f
List available metrics
.RE
.RS 4
.TP 4
\fBversion\f
Display version
.RE
.RE
.IP "\fB\-p\fR, \fB\-\-print\fR"
Print known items and exit.
For each item either generic defaults are used, or specific defaults for testing are supplied.
These defaults are listed in square brackets as item key parameters.
Returned values are enclosed in square brackets and prefixed with the type of the returned value, separated by a pipe character.
For user parameters type is always \fBt\fR, as the agent can not determine all possible return values.
Items, displayed as working, are not guaranteed to work from the Zabbix server or zabbix_get when querying a running agent daemon as permissions or environment may be different.
Returned value types are:
.RS 4
.TP 4
d
Number with a decimal part.
.RE
.RS 4
.TP 4
m
Not supported.
This could be caused by querying an item that only works in the active mode like a log monitoring item or an item that requires multiple collected values.
Permission issues or incorrect user parameters could also result in the not supported state.
.RE
.RS 4
.TP 4
s
Text.
Maximum length not limited.
.RE
.RS 4
.TP 4
t
Text.
Same as \fBs\fR.
.RE
.RS 4
.TP 4
u
Unsigned integer.
.RE
.IP "\fB\-t\fR, \fB\-\-test\fR \fIitem\-key\fR"
Test single item and exit.
See \fB\-\-print\fR for output description.
.IP "\fB\-h\fR, \fB\-\-help\fR"
Display this help and exit.
.IP "\fB\-V\fR, \fB\-\-version\fR"
Output version information and exit.
.SH FILES
.TP
.I /usr/local/etc/zabbix_agent2.conf
Default location of Zabbix agent 2 configuration file (if not modified during compile time).
.SH "SEE ALSO"
.UR https://www.zabbix.com/manuals
Documentation
.UE
.PP
.BR zabbix_agentd (8),
.BR zabbix_get (8),
.BR zabbix_proxy (8),
.BR zabbix_sender (8),
.BR zabbix_server (8),
.BR zabbix_js (8)
.SH AUTHOR
Zabbix LLC


ChangeLog
Changes for 4.4.5rc1

New features:
..F....... [ZBXNEXT-5473] reworked template linking with Multiselect and loading macro list with AJAX (averza, Ivo, vmurzins)

Bug fixes:
..F....... [ZBX-16481] fixed spelling issues in the code (miks)
.......... [ZBX-1977] reverted fix for axis labels calculation (rlataria)
..F....... [ZBX-16894] fixed dynamic rows losing old input at form error (talbergs)
..F....... [ZBX-17019] fixed aria-live message partially visible beneath multiselect controls (averza)
........S. [ZBX-16966] fixed not operator in correlation function expression (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.4.4

4.4.4rc2 was released as 4.4.4 without any changes

--------------------------------------------------------------------------------
Changes for 4.4.4rc2

Bug fixes:
....I....T [ZBXNEXT-5386] added missing webhook parameters to default database data (Sasha)

--------------------------------------------------------------------------------
Changes for 4.4.4rc1

New features:
.........T [ZBXNEXT-5623] added Pushover webhook integration (vitaly)
.........T [ZBXNEXT-5591] added media "Mattermost" (afayantsev)
.........T [ZBXNEXT-5586] added media "Opsgenie" (mchudinov)
...G...... [ZBXNEXT-5609] added agent2 Windows build support (wiper)
A.F....... [ZBXNEXT-5271] implemented 'delete missing' option for imported template linkages (miks)
.........T [ZBXNEXT-5600] updated templates to internal version v0.34 (vitaly)
.........T [ZBXNEXT-5572] added windows services template, updated windows template objects descriptions, added missing user macro (vitaly)

Bug fixes:
..F....... [ZBX-17101] fixed url validation before output (miks)
...G...... [ZBX-17070] fixed agent2 plugin configuration not being called if no plugin specific parameters are set (wiper)
........S. [ZBX-17062] fixed LLD macro substitution when PostgrSQL is used (viktors)
.......PS. [ZBX-16929] removed redundant code block (mabele)
..F....... [ZBX-16804] added missing get parameters in availability report (rlataria)
..F....... [ZBX-16286] fixed graph widget dimensions errors on high DPI screens (agriscenko)
..F....... [ZBX-16848] fixed fatal error occurring in user profile and user edit forms when PHP FileInfo extension does not exist (Ivo)
..F....... [ZBX-16684] fixed incorrect double quotes in history plain text view (Ivo)
..F....... [ZBX-16918] fixed widget saving with slow internet connection (rlataria)
A......... [ZBX-16903] fixed validation of "interfaceid" field for HTTP agent items with large IDs (Ivo)
..F....... [ZBX-16751] fixed support for php 7.4 (talbergs)
..F....... [ZBX-16879] changed condition description message for tag value in actions and event correlations (vasilijs)
........S. [ZBX-16459] fixed context-aware LLD macro expansion in JsonPath preprocessing (atumilovics)
..F....... [ZBX-16911] fixed multiselect searching suggests in template mass update (rlataria)
.......PS. [ZBX-9084] fixed high memory usage during startup (vso)
...G...... [ZBX-17058] fixed agent2 passive check timemouts exhausting plugin capacity (wiper)
...G...... [ZBX-16932] fixed build fail on netBSD (mabele)
..F....... [ZBX-15041] fixed request not being cancelled along with popup window in widgets (talbergs)
A.F....... [ZBX-16111] removed templateid from screen API output (vasilijs)
..F....... [ZBX-16919] fixed server check warning width in chrome (ashubin, rlataria)
..F....... [ZBX-16954] fixed graph widget scroll disappear (rlataria)
..F....... [ZBX-16995] fixed widget vertically stretch (averza, rlataria)
..F....... [ZBX-16978] fixed undefined offset in problem by severity widget (rlataria)
..F....... [ZBX-16944] fixed "Type of information" field in item form being marked as required when it is read-only (Ivo)
..F....... [ZBX-16912] fixed image map elements having a hand cursor when there is no context menu available (Ivo)
..F....... [ZBX-1977] fixed axis labels calculation (ashubin, rlataria)
..F....... [ZBX-16877] fixed long SNMP OID value goes beyond fields block (rlataria)
..F....... [ZBX-17034] updated Zabbix website links (averza)
..F....... [ZBX-16874] fixed map-type widget clipping in Internet Explorer (averza)
.......PS. [ZBX-16915] fixed memory leak (mabele)
........S. [ZBX-17008] improved Zabbix server performance when using maintenance (vso)
........S. [ZBX-16779] fixed infinite loop when writing export to file fails (mabele)
.........T [ZBX-16540] splitted host_resources templates into 3 subtemplates: for CPU, memory and storage (vitaly)
..F....... [ZBX-16852] fixed graph widget multiselect collapse (ashubin, rlataria)
..F....... [ZBX-16649] fixed flexible textarea behavior (averza, vasilijs)

--------------------------------------------------------------------------------
Changes for 4.4.3

Bug fixes:
...G...PS. [ZBX-16988] fixed agent memory leak (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.4.2

4.4.2rc2 was released as 4.4.2 without any changes

--------------------------------------------------------------------------------
Changes for 4.4.2rc2

Bug fixes:
........S. [ZBX-16955] fixed performance of history syncers and timer processes by not locking each other when suppressing events (vso)

--------------------------------------------------------------------------------
Changes for 4.4.2rc1

New features:
.........T [ZBXNEXT-5493,ZBX-16540] moved LLD rules from parent templates to linked templates for Module HOST-RESOURCES-MIB SNMPv2, Module Interfaces Windows SNMPv2, Net Arista SNMPv2, OS Windows SNMPv2 (vitaly)
...G...... [ZBXNEXT-5570] changed agent2 plugin configuration, moved MaxLinesPerSecond, EnableRemoteCommands, LogRemoteCommands from global to plugin configuration (wiper)
..F....... [ZBX-16871] fixed not setting the default values of multiselects on initial load (averza)
..F....... [ZBX-16776] reworked custom item select to multiselect (rlataria)

src/go/cmd/zabbix_agent2/zabbix_agent2.go
/*
** Zabbix
** Copyright (C) 2001-2020 Zabbix SIA
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/

package main

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	_ "zabbix.com/plugins"

	"zabbix.com/internal/agent"
	"zabbix.com/internal/agent/remotecontrol"
	"zabbix.com/internal/agent/scheduler"
	"zabbix.com/internal/agent/serverconnector"
	"zabbix.com/internal/agent/serverlistener"
	"zabbix.com/internal/agent/statuslistener"
	"zabbix.com/internal/monitor"
	"zabbix.com/pkg/conf"
	"zabbix.com/pkg/log"
	"zabbix.com/pkg/pidfile"
	"zabbix.com/pkg/tls"
	"zabbix.com/pkg/version"
	"zabbix.com/pkg/zbxlib"
)

const hostNameLen = 128

func configDefault(taskManager scheduler.Scheduler, o *agent.AgentOptions) error {
	var err error

	if len(o.Hostname) == 0 {
		var hostnameItem string

		if len(o.HostnameItem) == 0 {
			hostnameItem = "system.hostname"
		} else {
			hostnameItem = o.HostnameItem
		}

		o.Hostname, err = taskManager.PerformTask(hostnameItem, time.Second*time.Duration(o.Timeout))
		if err != nil {
			if len(o.HostnameItem) == 0 {
				return fmt.Errorf("cannot get system hostname using \"%s\" item as default for \"HostnameItem\" configuration parameter: %s", hostnameItem, err.Error())
			}

			return fmt.Errorf("cannot get system hostname using \"%s\" item specified by \"HostnameItem\" configuration parameter: %s", hostnameItem, err.Error())
		}

		if len(o.Hostname) == 0 {
			return fmt.Errorf("cannot get system hostname using \"%s\" item specified by \"HostnameItem\" configuration parameter: value is empty", hostnameItem)
		}

		if len(o.Hostname) > hostNameLen {
			o.Hostname = o.Hostname[:hostNameLen]
			log.Warningf("the returned value of \"%s\" item specified by \"HostnameItem\" configuration parameter is too long, using first %d characters", hostnameItem, hostNameLen)
		}

		if err = agent.CheckHostname(o.Hostname); nil != err {
			return fmt.Errorf("cannot get system hostname using \"%s\" item specified by \"HostnameItem\" configuration parameter: %s", hostnameItem, err.Error())
		}
	} else {
		if len(o.HostnameItem) != 0 {
			log.Warningf("both \"Hostname\" and \"HostnameItem\" configuration parameter defined, using \"Hostname\"")
		}
	}

	return nil
}

func configValidate(o *agent.AgentOptions) {
	var err error
	var exit bool = false

	if len(o.Hostname) > hostNameLen {
		log.Critf("invalid \"Hostname\" configuration parameter: configuration parameter cannot be longer than %d characters", hostNameLen)
		exit = true
	}
	if err = agent.CheckHostname(o.Hostname); nil != err {
		log.Critf("invalid \"Hostname\" configuration parameter: %s", err.Error())
		exit = true
	}
	if len(o.HostMetadata) > 0 && len(o.HostMetadata) > serverconnector.HostMetadataLen {
		log.Critf("the value of \"HostMetadata\" configuration parameter cannot be longer than %d characters", serverconnector.HostMetadataLen)
		exit = true
	}
	if len(o.HostInterface) > serverconnector.HostInterfaceLen {
		log.Critf("the value of \"HostInterface\" configuration parameter cannot be longer than %d characters", serverconnector.HostInterfaceLen)
		exit = true
	}
	if exit {
		os.Exit(1)
	}
}

var manager *scheduler.Manager
var listeners []*serverlistener.ServerListener
var serverConnectors []*serverconnector.Connector

func processLoglevelCommand(c *remotecontrol.Client, params []string) (err error) {
	if len(params) != 2 {
		return errors.New("No 'loglevel' parameter specified")
	}
	switch params[1] {
	case "increase":
		if log.IncreaseLogLevel() {
			message := fmt.Sprintf("Increased log level to %s", log.Level())
			log.Infof(message)
			err = c.Reply(message)
		} else {
			err = fmt.Errorf("Cannot increase log level above %s", log.Level())
			log.Infof(err.Error())
		}
	case "decrease":
		if log.DecreaseLogLevel() {
			message := fmt.Sprintf("Decreased log level to %s", log.Level())
			log.Infof(message)
			err = c.Reply(message)
		} else {
			err = fmt.Errorf("Cannot decrease log level below %s", log.Level())
			log.Infof(err.Error())
		}
	default:
		return errors.New("Invalid 'loglevel' parameter")
	}
	return
}

func processMetricsCommand(c *remotecontrol.Client, params []string) (err error) {
	data := manager.Query("metrics")
	return c.Reply(data)
}

func processVersionCommand(c *remotecontrol.Client, params []string) (err error) {
	data := version.Long()
	return c.Reply(data)
}

func processHelpCommand(c *remotecontrol.Client, params []string) (err error) {
	help := `Remote control interface, available commands:
	loglevel increase - Increase log level
	loglevel decrease - Decrease log level
	metrics - List available metrics
	version - Display Agent version
	help - Display this help message`
	return c.Reply(help)
}

func processRemoteCommand(c *remotecontrol.Client) (err error) {
	params := strings.Fields(c.Request())
	if len(params) == 0 {
		return errors.New("Empty command")
	}
	switch params[0] {
	case "loglevel":
		err = processLoglevelCommand(c, params)
	case "help":
		err = processHelpCommand(c, params)
	case "metrics":
		err = processMetricsCommand(c, params)
	case "version":
		err = processVersionCommand(c, params)
	default:
		return errors.New("Unknown command")
	}
	return
}

var pidFile *pidfile.File

func run() (err error) {
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	var control *remotecontrol.Conn
	if control, err = remotecontrol.New(agent.Options.ControlSocket); err != nil {
		return
	}
	control.Start()

loop:
	for {
		select {
		case sig := <-sigs:
			switch sig {
			case syscall.SIGINT, syscall.SIGTERM:
				break loop
			}
		case client := <-control.Client():
			if rerr := processRemoteCommand(client); rerr != nil {
				if rerr = client.Reply("error: " + rerr.Error()); rerr != nil {
					log.Warningf("cannot reply to remote command: %s", rerr)
				}
			}
			client.Close()
		}
	}
	control.Stop()
	return nil
}

var confDefault string

func main() {
	var confFlag string
	const (
		confDescription = "Path to the configuration file"
	)
	flag.StringVar(&confFlag, "config", confDefault, confDescription)
	flag.StringVar(&confFlag, "c", confDefault, confDescription+" (shorthand)")

	var foregroundFlag bool
	const (
		foregroundDefault     = true
		foregroundDescription = "Run Zabbix agent in foreground"
	)
	flag.BoolVar(&foregroundFlag, "foreground", foregroundDefault, foregroundDescription)
	flag.BoolVar(&foregroundFlag, "f", foregroundDefault, foregroundDescription+" (shorthand)")

	var testFlag string
	const (
		testDefault     = ""
		testDescription = "Test specified item and exit"
	)
	flag.StringVar(&testFlag, "test", testDefault, testDescription)
	flag.StringVar(&testFlag, "t", testDefault, testDescription+" (shorthand)")

	var printFlag bool
	const (
		printDefault     = false
		printDescription = "Print known items and exit"
	)
	flag.BoolVar(&printFlag, "print", printDefault, printDescription)
	flag.BoolVar(&printFlag, "p", printDefault, printDescription+" (shorthand)")

	var versionFlag bool
	const (
		versionDefault     = false
		versionDescription = "Print program version and exit"
	)
	flag.BoolVar(&versionFlag, "version", versionDefault, versionDescription)
	flag.BoolVar(&versionFlag, "V", versionDefault, versionDescription+" (shorthand)")

	var remoteCommand string
	const (
		remoteDefault     = ""
		remoteDescription = "Perform administrative functions (send 'help' for available commands)"
	)
	flag.StringVar(&remoteCommand, "R", remoteDefault, remoteDescription)

	flag.Parse()

	var argConfig, argTest, argPrint, argVersion bool

	// Need to manually check if the flag was specified, as default flag package
	// does not offer automatic detection. Consider using third party package.
	flag.Visit(func(f *flag.Flag) {
		switch f.Name {
		case "c", "config":
			argConfig = true
		case "t", "test":
			argTest = true
		case "p", "print":
			argPrint = true
		case "V", "version":
			argVersion = true
		}
	})

	if argVersion {
		version.Display()
		os.Exit(0)
	}

	if err := conf.Load(confFlag, &agent.Options); err != nil {
		if argConfig || !(argTest || argPrint) {
			fmt.Fprintf(os.Stderr, "%s\n", err.Error())
			os.Exit(1)
		}
		// create default configuration for testing options
		if !argConfig {
			_ = conf.Unmarshal([]byte{}, &agent.Options)
		}
	}

	if argTest || argPrint {
		if err := log.Open(log.Console, log.Warning, "", 0); err != nil {
			fmt.Fprintf(os.Stderr, "Cannot initialize logger: %s\n", err.Error())
			os.Exit(1)
		}

		if argTest {
			if err := agent.CheckMetric(testFlag); err != nil {
				os.Exit(1)
			}
		} else {
			agent.CheckMetrics()
		}

		os.Exit(0)
	}

	if remoteCommand != "" {
		if agent.Options.ControlSocket == "" {
			fmt.Fprintf(os.Stderr, "Cannot send remote command: ControlSocket configuration parameter is not defined\n")
			os.Exit(0)
		}

		if reply, err := remotecontrol.SendCommand(agent.Options.ControlSocket, remoteCommand); err != nil {
			fmt.Fprintf(os.Stderr, "Cannot send remote command: %s\n", err)
		} else {
			fmt.Fprintf(os.Stdout, "%s\n", reply)
		}
		os.Exit(0)
	}

	var logType, logLevel int
	switch agent.Options.LogType {
	case "console":
		logType = log.Console
	case "file":
		logType = log.File
	}
	switch agent.Options.DebugLevel {
	case 0:
		logLevel = log.Info
	case 1:
		logLevel = log.Crit
	case 2:
		logLevel = log.Err
	case 3:
		logLevel = log.Warning
	case 4:
		logLevel = log.Debug
	case 5:
		logLevel = log.Trace
	}

	if err := log.Open(logType, logLevel, agent.Options.LogFile, agent.Options.LogFileSize); err != nil {
		fmt.Fprintf(os.Stderr, "Cannot initialize logger: %s\n", err.Error())
		os.Exit(1)
	}

	zbxlib.SetLogLevel(logLevel)

	greeting := fmt.Sprintf("Starting Zabbix Agent 2 [%s]. (%s)", agent.Options.Hostname, version.Long())
	log.Infof(greeting)

	addresses, err := serverconnector.ParseServerActive()
	if err != nil {
		log.Critf("%s", err)
		os.Exit(1)
	}

	if tlsConfig, err := agent.GetTLSConfig(&agent.Options); err != nil {
		log.Critf("cannot use encryption configuration: %s", err)
		os.Exit(1)
	} else {
		if tlsConfig != nil {
			if err := tls.Init(tlsConfig); err != nil {
				log.Critf("cannot configure encryption: %s", err)
				os.Exit(1)
			}
		}
	}

	if pidFile, err = pidfile.New(agent.Options.PidFile); err != nil {
		log.Critf(err.Error())
		os.Exit(1)
	}
	defer pidFile.Delete()

	if foregroundFlag {
		if agent.Options.LogType != "console" {
			fmt.Println(greeting)
		}
		fmt.Println("Press Ctrl+C to exit.")
	}

	log.Infof("using configuration file: %s", confFlag)

	if err = agent.InitUserParameterPlugin(agent.Options.UserParameter, agent.Options.UnsafeUserParameters); err != nil {
		log.Critf("cannot initialize user parameters: %s", err)
		os.Exit(1)
	}

	if manager, err = scheduler.NewManager(&agent.Options); err != nil {
		log.Critf("cannot create scheduling manager: %s", err)
		os.Exit(1)
	}

	// replacement of deprecated StartAgents
	if 0 != len(agent.Options.Server) {
		var listenIPs []string
		if listenIPs, err = serverlistener.ParseListenIP(&agent.Options); nil != err {
			log.Critf(err.Error())
			os.Exit(1)
		}
		for i := 0; i < len(listenIPs); i++ {
			listener := serverlistener.New(i, manager, listenIPs[i], &agent.Options)
			listeners = append(listeners, listener)
		}
	}

	manager.Start()

	if err = configDefault(manager, &agent.Options); err != nil {
		log.Critf("cannot process configuration: %s", err)
		os.Exit(1)
	}

	configValidate(&agent.Options)

	serverConnectors = make([]*serverconnector.Connector, len(addresses))

	for i := 0; i < len(serverConnectors); i++ {
		if serverConnectors[i], err = serverconnector.New(manager, addresses[i], &agent.Options); err != nil {
			log.Critf("cannot create server connector: %s", err)
			os.Exit(1)
		}
		serverConnectors[i].Start()
	}

	for _, listener := range listeners {
		if err = listener.Start(); nil != err {
			log.Critf("cannot start server listener: %s", err)
			os.Exit(1)
		}
	}

	if agent.Options.StatusPort != 0 {
		if err = statuslistener.Start(manager, confFlag); err != nil {
			log.Critf("cannot start HTTP listener: %s", err)
			os.Exit(1)
		}
	}

	if err == nil {
		err = run()
	}
	if err != nil {
		log.Errf("cannot start agent: %s", err.Error())
	}

	if agent.Options.StatusPort != 0 {
		statuslistener.Stop()
	}

	for _, listener := range listeners {
		listener.Stop()
	}
	for i := 0; i < len(serverConnectors); i++ {
		serverConnectors[i].Stop()
	}
	manager.Stop()

	monitor.Wait()

	farewell := fmt.Sprintf("Zabbix Agent 2 stopped. (%s)", version.Long())
	log.Infof(farewell)

	if foregroundFlag {
		if agent.Options.LogType != "console" {
			fmt.Println(farewell)
		}
		fmt.Println("Press Ctrl+C to exit.")
	}
}
